#Большие данные
#Отчет. Работа с фреймами данных. Библиотека Pandas. Экспертные задания 1
#Подготовил: Подрядчиков Владимир Валерьевич, группа 02421-ДБ

# <center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)

In [ ]:
import pandas as pd
import numpy as np

#### Task 1.
Задание:

1. Создайте Series из последовательности 15 значений, равномерно разбивающих отрезок [0, 20] (воспользуйтесь функцией linspace)
2. Определите отношение элементов полученной серии к их предыдущим элементам (\*).
3. В результате необходимо получить среднее полученного вектора, оставив в нём только те значения, которые не более чем 1.5 (\**).

Выберите из ответов тот, который максимально близок к полученному (c точки зрения абсолютной разницы).

#### Варианты ответов:
    - 1) 1.24
    - 2) 1.18
    - 3) 0.71
    - 4) 1.13


##### Пояснения:
(\*) Если было бы необходимо найти последовательность из 3-х значений,  
равномерно разбивающих отрезок [0,1], то это были бы значения [0, 0.5, 1].

(\**) Если был бы дан список элементов a = [1,2,3,12], отношения элементов к предыдущим будут равны [NaN, 2, 1.5, 4].  
А на последнем этапе в таком примере останется только [1.5] и среднее значение будет также 1.5.

In [ ]:
series = pd.Series(np.linspace(0, 20, 15))
ratios = series / series.shift(1)
ratios_filter = ratios[ratios <= 1.5]
result = ratios_filter.mean()

print(result)

1.181677812927813


Ответ: **Вариант 2) 1.18**

#### Task 2.
Выберите все верные ответы касательно следующих 3-х Series:
    - pd.Series('abcde'); (1)
    - pd.Series(['abcde']); (2)
    - pd.Series(list('abcde')); (3)
    - pd.Series("abcde"); (4)
    
##### Пояснения:
(\*)
- функция list: в строке каждый символ - это отдельный элемент для list
- квадратные скобки: в квадратных скобках списку передается множество элементов по отдельности через запятую


#### Вопросы:
    - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк
    - 2) Серия (2) совпадает с серией (3), так как в каждом из случаев серия создаётся из списка символов
    - 3) Серия (1) не совпадает с серией (4), так как в (4) используются двойные кавычки "" вместо одинарных ''

In [ ]:
#Инициализируем Series элементы
first = pd.Series('abcde')
second = pd.Series(['abcde'])
third = pd.Series(list('abcde'))
forth = pd.Series("abcde")

In [ ]:
print(first)
print(second)
print(third)
print(forth)

0    abcde
dtype: object
0    abcde
dtype: object
0    a
1    b
2    c
3    d
4    e
dtype: object
0    abcde
dtype: object


В результате видим, что объекты 1,2,4 содержат один элемент abcde, только объект 3 содержит 5 символов по отдельности.

**Верно утверждение 1)**

#### Task 3.

По клиенту получены зашумленные данные (объект s типа Series) по его транзакциям.

Для заданного ниже объекта s проделайте следующее:
1. Создайте новый Series, значения которого совпадают со значениями s, а индексы - целочисленные значения от 2 до 12, не включая 12.
2. Выберите из s элементы с индексами 3 и 5, после чего просуммируйте их, сохранив результат (1).
3. Выберите из s только целочисленные элементы и вычислите их дисперсию (2). (\*)

Все полученные результаты округлите до 2-х знаков после запятой.

#### Выберите все верные пункты:
    - 1) Ответ (1) - 642.52
    - 2) Ответ (1) - 91.78
    - 3) Ответ (1) - не может быть определён
    - 4) Ответ (2) - 57591.19
    - 5) Ответ (2) - 210.12
    - 6) Ответ (2) - не может быть определён

##### Пояснения:
(\*) Целочисленные значения - значения, имеющие тип int.
- Дисперсия рассчитывается с помощью функции из библиотеки numpy: np.var( , ddof=0) или встроенной в python функции: .var(ddof=1)

In [ ]:
s = pd.Series(data=['1', 2, 3.1, 'hi!', 5, -512, 12.42, 'sber', 10.10, 98],
              index=range(6, 26, 2))

new = pd.Series(data=s.values, index=range(2, 12))

#sum_1 = s.loc[[3, 5]].astype(float).sum()

nums = pd.to_numeric(s, errors='coerce').dropna()
integers = nums[nums.astype(int) == nums]
disp_2 = round(np.var(integers , ddof=1), 2)

print(integers)
print(disp_2)


6       1.0
8       2.0
14      5.0
16   -512.0
24     98.0
dtype: float64
59702.7


В s вообще нет элементов с индексами 3 и 5. Поэтому нельзя просуммировать, выдаст ошибку. Если нужно было просуммировать значения из new, то там это тоже нельзя сделать, так как один из элементов типа String.

Для второго условия сначала перевели элементы в числа, если это возможно. Искочили то, что перевести не удалось и дальше взяли те, что при переводе в целые остались самими собой.
Дисперсия целочисленных элементов будет равна 59702,2

Ответ: 3) и 4)(наиболее близок к посчитанному значению)

#### Task 4.
1. Сгенерируйте Series из 100 значений нормально распределённой СВ (np.random.normal с дефолтными параметрами - нулевым средним и единичной дисперсией).
2. Возведите каждое значение серии в 3 степень, а значения индекса увеличьте в 3 раза.
3. Ответьте на следующие вопросы через запятую (без пробелов) (\*)
    - 1) Выведите сумму элементов, строго меньших 2.6, имеющих нечётные значения индекса.
    - 2) Выведите количество значений серии меньше нуля.

#### Пояснения:
(\*) Если получились ответы 3, 4.32, то в форму необходимо внести их в виде "3,4.32". То есть вещественные числа необходимо разделять точками. Не забудьте про фиксированный seed (его менять не нужно)!
 - Определенное значение seed нужно, чтобы ответы у всех выполняющих
это задание были одинаковые и их можно было проверить (так как генерируются
одинаковые series).
 - Следует внимательнее использовать [ ] для выбора данных по нескольким условиям:
либо выбирать данные последовательно, либо сразу по нескольким условиям,
но через оператор &. Отличие оператора and от оператора &: and - выводит
последнее проверенное значение, & - выводит пересечение значений.
Пример: s[ _  &  _ ].sum()

In [ ]:
np.random.seed(242)

s = pd.Series(np.random.normal(size=100))
s = s**3
s.index = s.index * 3

sum_odd= s[(s < 2.6) & (s.index % 2 != 0)].sum()

less_0 = s[s < 0]
len_less_0 = len(less_0)

print(sum_odd)
print(len_less_0)


-15.150748101821666
51


Ответ:   
1) -15,15  
2) 51

### Информация для последующих заданий
- Для всех последующих заданий будем использовать обезличенные транзакционные банковские данные. Для этого считайте в переменные tr_mcc_codes, tr_types, transactions и gender_train из одноимённых таблиц из папки data. Для таблицы transactions используйте только первые n=1000000 строк. Обратите внимание на разделители внутри каждого из файлов - они могут различаться!

In [ ]:
tr_mcc_codes = pd.read_table('tr_mcc_codes.csv', sep = ';')
tr_types = pd.read_table('tr_types.csv', sep = ';')
transactions = pd.read_table('transactions.csv', nrows = 1000000, sep = ',')
gender_train = pd.read_table('gender_train.csv', sep = ',')

# Описание данных
### Таблица ```transactions.csv```
##### Описание
Таблица содержит историю транзакций клиентов банка за один год и три месяца.

##### Формат данных

```
customer_id,tr_datetime,mcc_code,tr_type,amount,term_id
111111,15 01:40:52,1111,1000,-5224,111111
111112,15 15:18:32,3333,2000,-100,11122233
...
```
##### Описание полей

 - ```customer_id``` — идентификатор клиента;
 - ```tr_datetime``` — день и время совершения транзакции (дни нумеруются с начала данных);
 - ```mcc_code``` — mcc-код транзакции;
 - ```tr_type``` — тип транзакции;
 - ```amount``` — сумма транзакции в условных единицах со знаком; ```+``` — начисление средств клиенту (приходная транзакция), ```-``` — списание средств (расходная транзакция);
 - ```term_id``` — идентификатор терминала;


### Таблица ```gender_train.csv```

##### Описание
Данная таблица содержит информацию по полу для части клиентов, для которых он известен. Для остальных клиентов пол неизвестен.

##### Формат данных
```
customer_id,gender
111111,0
111112,1
...
```

##### Описание полей
 - ```customer_id``` — идентификатор клиента;
 - ```gender``` — пол клиента;

### Таблица ```tr_mcc_codes.csv```

##### Описание
Данная таблица содержит описание mcc-кодов транзакций.

##### Формат данных
```
mcc_code;mcc_description
1000;словесное описание mcc-кода 1000
2000;словесное описание mcc-кода 2000
...
```

##### Описание полей
 - ```mcc_code``` – mcc-код транзакции;
 - ```mcc_description``` — описание mcc-кода транзакции.

### Таблица ```tr_types.csv```

##### Описание
Данная таблица содержит описание типов транзакций.

##### Формат данных
```
tr_type;tr_description
1000;словесное описание типа транзакции 1000
2000;словесное описание типа транзакции 2000
...
```

##### Описание полей
 - ```tr_type``` – тип транзакции;
 - ```tr_description``` — описание типа транзакции;

#### Task 5.
1. В tr_types выберите произвольные 100 строк с помощью метода sample (указав при этом random_seed равный 242)
2. В полученной на предыдущем этапе подвыборке найдите долю наблюдений (стобец tr_description), в которой содержится подстрока 'плата' (в любом регистре). (\*)

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Строки "ПлатА за аренду", "ПлатАза аренду", "ПЛАТА" удовлетворяют условию, так как будучи переведёнными в нижний регистр содержат подстроку "плата".

In [ ]:
str_100 = tr_types.sample(n=100, random_state=242)

count_plata = str_100['tr_description'].str.lower().str.contains('плата').sum()

frac_plata = count_plata / len(str_100)

print("{:.2f}".format(frac_plata))

0.26


Ответ: 0.26

#### Task 6.
1. Для поля tr_type датафрейма transactions посчитайте частоту встречаемости всех типов транзакций tr_type в transactions.
2. Из перечисленных вариантов выберите те, которые попали в топ-5 транзакций по частоте встречаемости.

#### Выберите все верные пункты:
    - 1) Выдача наличных в АТМ Сбербанк России
    - 2) Комиссия за обслуживание ссудного счета
    - 3) Списание по требованию
    - 4) Оплата услуги. Банкоматы СБ РФ
    - 5) Погашение кредита (в пределах одного филиала)
    - 6) Покупка. POS ТУ СБ РФ

In [ ]:
tr_type_count = transactions['tr_type'].value_counts()

top_5_code = tr_type_count.head(5)

top_5_desc = pd.merge(pd.DataFrame(top_5_code), tr_types, left_index=True, right_on='tr_type')

print(top_5_desc['tr_description'])

98                                 Покупка. POS ТУ СБ РФ
106                Выдача наличных в АТМ Сбербанк России
101                               Покупка. POS ТУ Россия
99                        Оплата услуги. Банкоматы СБ РФ
59     Перевод на карту (с карты) через Мобильный бан...
Name: tr_description, dtype: object


Ответ:  
1) Выдача наличных в АТМ Сбербанк России  
4) Оплата услуги. Банкоматы СБ РФ  
6) Покупка. POS ТУ СБ РФ

#### Task 7.
1. В датафрейме transactions задайте столбец  customer_id в качестве индекса.
2. Выделите клиента с максимальной суммой транзакции (то есть с максимальным приходом на карту). (\*)
3. Найдите у него наиболее часто встречающийся модуль суммы приходов/расходов. (\**)


#### Выберите все верные пункты:
    - 1) 1122957.89
    - 2) 15721.41
    - 3) 22459.16
    - 4) 13475494.63
    - 5) 107407.78
    - 6) 65019.26

##### Пояснения:
(\*) Если у клиента были транзакции [-10000, 10, 0, -10], то максимумом будет являться значение 10. <br>
(\**) Если у клиента были транзакции [-10000, 10, 0, -10], то наиболее встречающийся модуль суммы транзакций равен 10, и встретился он 2 раза.

In [ ]:
transactions = pd.read_table('transactions.csv', index_col='customer_id', nrows = 1000000, sep = ',')

id_max = transactions['amount'].idxmax()
trans_max = transactions['amount'].max()

trans_grouped2 = transactions.loc[id_max, 'amount'].abs().value_counts()
abs_max = trans_grouped2.idxmax()

print(id_max)
print(trans_max)
print(abs_max)


26782922
145984525.17
22459.16


Id клиента с максимальной суммой транзакции: 26782922. Его максимальная транзакция 145984525.17

Ответ: Наиболее часто встречающийся модуль суммы приходов/расходов
3) 22459.16

#### Task 8.
1. Найдите максимальную разницу между медианами суммы транзакций, посчитанными при заданных ниже условиях по полю amount из таблицы transactions (\*):
    - Медиана суммы транзакций
    - Медиана суммы транзакций по тем строкам, которые ни в одном из своих столбцов не содержат пустые значения
    - Медиана суммы транзакций по строкам, отсортированным по полю amount в порядке возрастания, и из которых удалены дублирующиеся по столбцам [mcc_code, tr_type] строки, причём при удалении соответстующих дублей остаются только последние из дублирующихся строк (keep='last')

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Для вычисления максимальной разницы между значениями списка можно использовать функцию np.ptp  
(\**) Если в результате получились значения [1,3,5], то макимальная разница между ними 4 == 5-1.

In [ ]:
import warnings
warnings.filterwarnings('ignore')


med_sum = transactions['amount'].median()

transactions_noNan = transactions.dropna()
med_sum_withoutNan = transactions_noNan.median()

sorted = transactions.sort_values('amount')
drop_dup = sorted.drop_duplicates(subset=['mcc_code', 'tr_type'], keep='last')
med_sum_sort_nodup = drop_dup.median()

print(med_sum)
print(med_sum_withoutNan['amount'])
print(med_sum_sort_nodup['amount'])

max_diff = np.max([abs(med_sum - med_sum_withoutNan['amount']), abs(med_sum - med_sum_sort_nodup['amount']), abs(med_sum_sort_nodup['amount'] - med_sum_withoutNan['amount'])])

print("{:.2f}".format(max_diff))


-5614.79
-5596.6
-920.83
4693.96


Ответ:  
Медиана 1: -5614.79  
Медиана 2: -5596,6  
Медиана 3: -920,83  
Наибольшая разница - 4693.96